compiler-core/src/erlang.rs

```diff
@@ -512,6 +512,37 @@ fn tuple<'a>(elems: impl IntoIterator<Item = Document<'a>>) -> Document<'a> {
         .group()
 }
 
+fn const_string_concatenate<'a>(
+    left: &'a TypedConstant,
+    right: &'a TypedConstant,
+    env: &mut Env<'a>,
+) -> Document<'a> {
+    let left = const_string_concatenate_argument(left, env);
+    let right = const_string_concatenate_argument(right, env);
+    bit_array([left, right])
+}
+
+fn const_string_concatenate_argument<'a>(
+    value: &'a TypedConstant,
+    env: &mut Env<'a>,
+) -> Document<'a> {
+    match value {
+        Constant::String { value, .. } => docvec!['"', string_inner(value), "\"/utf8"],
+
+        Constant::Var { constructor: Some(constructor), .. } => {
+            match &constructor.variant {
+                ValueConstructorVariant::ModuleConstant { literal: Constant::String { value, ..}, .. } => docvec!['"', string_inner(value), "\"/utf8"],
+                ValueConstructorVariant::ModuleConstant { literal: concat_value @ Constant::StringConcatenation { .. }, .. } => docvec![const_inline(concat_value, env), "/binary"],
+                _ => panic!("Constant string concatenation argument can only be a string or const var at code generation time"),
+            }
+        }
+
+        Constant::StringConcatenation { .. } => docvec![const_inline(value, env), "/binary"],
+
+        _ => panic!("Constant string concatenation argument can only be a string or const var at code generation time"),
```

Comment:
Please remove these panics, code generation is not the place to do type checking, and other variants could be suitable in future. Calling `const_inline` on whatever the `Constant::*` is is good.